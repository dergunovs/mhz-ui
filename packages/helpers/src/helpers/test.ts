import { createApp, defineComponent } from 'vue';
import type { RootNode, TemplateChildNode } from '@vue/compiler-core';
import { createRouter, createWebHistory } from 'vue-router';

const template = '<template><div></div></template>';

export function dataTest(value: string): string {
  return `[data-test="${value}"]`;
}

export function removeDataTest(node: RootNode | TemplateChildNode) {
  if (node.type === 1 /* NodeTypes.ELEMENT */ && 'props' in node) {
    node.props = node.props.filter((prop: { type: number; name: string }) =>
      prop.type === 6 ? prop.name !== 'data-test' : true
    );
  }
}

export async function wait(time?: number): Promise<void> {
  await new Promise((r) => {
    setTimeout(r, time || 10);
  });
}

export async function withSetup<T>(composable: () => Promise<T>) {
  const router = createRouter({
    history: createWebHistory(),
    routes: [
      { path: '/', name: 'Home', component: { template } },
      { path: '/login', name: 'Login', component: { template } },
    ],
  });

  let composableResult: Promise<T> | undefined;

  const component = defineComponent({
    template,
    setup() {
      composableResult = composable();
    },
  });

  const app = createApp(component);

  app.use(router);

  const wrapper = document.createElement('div');

  document.body.append(wrapper);
  app.mount(wrapper);

  await composableResult;
}
