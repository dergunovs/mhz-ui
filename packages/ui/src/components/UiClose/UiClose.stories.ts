import { Meta, StoryObj } from '@storybook/vue3';

import { html } from '@/utils';
import { UiClose } from '@/components';

const meta: Meta<typeof UiClose> = {
  component: UiClose,
  args: {},
  parameters: {
    docs: {
      description: {
        component: '',
      },
    },
  },
};

const argTypes = {};

type Story = StoryObj<typeof UiClose>;

export default meta;

export const Primary: Story = {
  render: (args) => ({
    components: { UiClose },
    setup: () => ({ args, argTypes }),

    template: html` <UiClose v-bind="args">{{DEFAULT_SLOT}}</UiClose>`,
  }),
  argTypes,
};
