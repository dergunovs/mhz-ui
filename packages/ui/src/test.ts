import { Component, ComponentPublicInstance } from 'vue';
import { shallowMount, VueWrapper } from '@vue/test-utils';
import { debounce } from 'perfect-debounce';

import { uiStubs } from './components/stubs/stubs';

export function wrapperFactory<T>(
  component: Component<T>,
  props?: Partial<ComponentPublicInstance<T>['$props']>,
  slots?: { default: string }
) {
  document.body.innerHTML = '<div id="app"></div>';

  return shallowMount(component, {
    global: {
      stubs: {
        ...uiStubs,
        RouterLink: { template: '<a><slot></slot></a>' },
        Teleport: { template: '<div><slot></slot></div>' },
      },
    },
    props: props as ComponentPublicInstance<T>['$props'],
    slots: slots as undefined,
    mocks: { debounce },
    attachTo: document.querySelector('#app') as HTMLElement,
  }) as VueWrapper<ComponentPublicInstance<T>>;
}
