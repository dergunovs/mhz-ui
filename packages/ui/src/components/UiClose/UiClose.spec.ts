import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { VueWrapper, enableAutoUnmount } from '@vue/test-utils';
import { dataTest } from 'mhz-helpers';

import UiClose from './UiClose.vue';

import { wrapperFactory } from '@/test';

const close = dataTest('ui-close');

let wrapper: VueWrapper<InstanceType<typeof UiClose>>;

beforeEach(() => {
  wrapper = wrapperFactory(UiClose);
});

enableAutoUnmount(afterEach);

describe('UiClose', async () => {
  it('exists', async () => {
    expect(wrapper.findComponent(UiClose)).toBeTruthy();
  });

  it('matches snapshot', async () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('sets style props', async () => {
    expect(wrapper.find(close).attributes('data-small')).toBe('false');
    expect(wrapper.find(close).attributes('data-delete')).toBe('false');

    await wrapper.setProps({ isSmall: true, isDelete: true });

    expect(wrapper.find(close).attributes('data-small')).toBe('true');
    expect(wrapper.find(close).attributes('data-delete')).toBe('true');
  });
});
