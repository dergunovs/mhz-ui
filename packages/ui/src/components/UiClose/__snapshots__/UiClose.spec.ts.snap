// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UiClose > matches snapshot 1`] = `
"<button class="_close_f6ab49" type="button" data-small="false" data-delete="false" data-test="ui-close">
  <div class="_line_f6ab49"></div>
  <div class="_line_f6ab49"></div>
</button>"
`;
